File: app/src/main/java/kr/zagros/shwan/moviemvvm/views/activities/MainActivity.java
package kr.zagros.shwan.moviemvvm.views.activities;


import androidx.lifecycle.ViewModelProviders;
import androidx.databinding.DataBindingUtil;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.widget.LinearLayout;
import android.widget.Toast;

import java.util.Objects;

import androidx.recyclerview.widget.RecyclerView;
import kr.zagros.shwan.moviemvvm.R;
import kr.zagros.shwan.moviemvvm.databinding.ActivityMain2Binding;

import kr.zagros.shwan.moviemvvm.viewmodels.MainVM;
import kr.zagros.shwan.moviemvvm.views.adapters.MoviesAdapter;

public class MainActivity extends AppCompatActivity implements RetryCallback {

    private MainVM mainVM;
    private ActivityMain2Binding binding;

    private MoviesAdapter moviesAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this,R.layout.activity_main2);
        mainVM = ViewModelProviders.of(this).get(MainVM.class);

        initNavigationDrawer();
        setupRecyclerView();

        setupObserversDataFromViewModel();

    }

    private void setupObserversDataFromViewModel() {
        mainVM.getResponseLiveData().observe(this, Movie -> moviesAdapter.submitList(Movie));
        mainVM.getNetworkStateLiveData().observe(this, networkState -> moviesAdapter.setMyNetworkState(networkState));
    }

    private void initNavigationDrawer() {
        setSupportActionBar(binding.toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        setSupportActionBar(binding.toolbar);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, binding.drawerLayout, binding.toolbar, R.string.Open, R.string.Close);
        toggle.syncState();
        binding.drawerLayout.addDrawerListener(toggle);
    }

    private void setupRecyclerView() {
        moviesAdapter = new MoviesAdapter(this, this);
        binding.content.list.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        binding.content.list.setAdapter(moviesAdapter);
    }

    @Override
    public void retry() {
        Toast.makeText(this, "retry", Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/kr/zagros/shwan/moviemvvm/views/activities/MainActivity.kt
package kr.zagros.shwan.moviemvvm.views.activities


import androidx.lifecycle.ViewModelProviders
import androidx.databinding.DataBindingUtil
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager

import android.widget.Toast
import androidx.lifecycle.Observer

import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.content_main.*
import kr.zagros.shwan.moviemvvm.Entities.NetworkState
import kr.zagros.shwan.moviemvvm.R
import kr.zagros.shwan.moviemvvm.databinding.ActivityMain2Binding

import kr.zagros.shwan.moviemvvm.viewmodels.MainVM
import kr.zagros.shwan.moviemvvm.views.adapters.MoviesAdapter

class MainActivity : AppCompatActivity(), RetryCallback {

    private var mainVM: MainVM? = null
    private var binding: ActivityMain2Binding? = null

    private var moviesAdapter: MoviesAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        mainVM = ViewModelProviders.of(this).get(MainVM::class.java)

        initNavigationDrawer()
        initRecyclerView()

        initObserversDataFromViewModel()

    }

    private fun initObserversDataFromViewModel() {
        mainVM!!.responseLiveData.observe(this, Observer { movie -> moviesAdapter!!.submitList(movie) })
        mainVM!!.networkStateLiveData.observe(this, Observer { networkState -> moviesAdapter!!.setMyNetworkState(networkState) })
        mainVM!!.firstDataState.observe(this, Observer { networkState ->
            run {
                if (networkState == NetworkState.LOADED)
                    main_progress.visibility = View.GONE
            }
        })
    }

    private fun initNavigationDrawer() {
        setSupportActionBar(binding!!.toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        setSupportActionBar(binding!!.toolbar)
        val toggle = ActionBarDrawerToggle(this, binding!!.drawerLayout, binding!!.toolbar, R.string.Open, R.string.Close)
        toggle.syncState()
        binding!!.drawerLayout.addDrawerListener(toggle)
    }

    private fun initRecyclerView() {
        moviesAdapter = MoviesAdapter(this, this)
        binding!!.content.list.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        binding!!.content.list.adapter = moviesAdapter
    }

    override fun retryGetData() {
        Toast.makeText(this, "retry", Toast.LENGTH_SHORT).show()
    }
}
